### ISSUE TITLE
[no-unnecessary-condition] code inside forEach not taken into account

### ISSUE THREAD

**Issue by yelhouti** [typescript-eslint/typescript-eslint#3270]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
let header: string | null = null;

event.headers.keys().forEach(entry => {
  if (entry === 'header') {
    header = event.headers.get(entry);
  }
});

if (header) { // error here
...
}
```

**Expected Result**

the condition to work ok

**Actual Result**

Unnecessary conditional, value is always falsy  @typescript-eslint/no-unnecessary-condition

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.20.0` |
| `@typescript-eslint/parser`        | `4.20.0` |
| `TypeScript`                       | `4.1.5` |
| `ESLint`                           | `7.23.0` |
| `node`                             | `14.16.0` |

**Comment by bradzacher** [comment]
I don't understand your code.
Why are you iterating over the entire headers keyset if you know you're looking for a key called `'header'`???
ESP considering your code does `if (entry === 'header') { get(entry) }`. This code **always** fetches the key `'header'`.

---

Regardless - this is just how TS sees the types. We work on top of TS's types.
TS's flow control does not understand that `forEach` is immediately called - so as far as TS is concerned - the variable `header` is never reassigned - meaning it's strictly typed as `null`.

![image](https://user-images.githubusercontent.com/7462525/113673006-46e5d080-966d-11eb-9b07-8976c3a06414.png)

Use better language constructs that TS can understand (like `for..of` loops), or write code that doesn't do as much wasteful iteration.

**Comment by yelhouti** [comment]
@bradzacher The code is just a simplified demo, but I get what you are saying, it is not your fault, so I guess I will create the issue somewhere else :D
